tasks/20231031-batch-relayer-v6/test/test.fork.ts

```diff
@@ -0,0 +1,191 @@
+    // need to use said ETH).
+
+    const ethStethTokens: Array<string> = (await vault.getPoolTokens(ETH_STETH_POOL)).tokens;
+    const stableWethIndex = ethStethTokens.findIndex((token) => token.toLowerCase() == WETH.toLowerCase());
+
+    const exitCalldata = library.interface.encodeFunctionData('exitPool', [
+      ETH_STETH_POOL,
+      0, // Even if this a Stable Pool, the Batch Relayer is unaware of their encodings and the Weighted Pool encoding
+      // happens to match here
+      relayer.address,
+      relayer.address,
+      {
+        assets: ethStethTokens,
+        minAmountsOut: ethStethTokens.map(() => 0),
+        // Note that we use the same input as before
+        userData: defaultAbiCoder.encode(['uint256', 'uint256', 'uint256'], [0, stakedBalance, stableWethIndex]),
+        toInternalBalance: true,
+      },
+      // Only store a chained reference for the WETH amount out, as the rest will be zero
+      [{ key: toChainedReference(42), index: stableWethIndex }],
+    ]);
+
+    // Join from WETH
+    const ethDaiTokens: Array<string> = (await vault.getPoolTokens(ETH_DAI_POOL)).tokens;
+    const ethDaiAmountsIn = ethDaiTokens.map((token) =>
+      token.toLowerCase() == WETH.toLowerCase() ? toChainedReference(42) : 0
+    );
+
+    const joinCalldata = library.interface.encodeFunctionData('joinPool', [
+      ETH_DAI_POOL,
+      0, // Weighted Pool
+      relayer.address,
+      relayer.address,
+      {
+        assets: ethDaiTokens,
+        maxAmountsIn: ethDaiTokens.map(() => MAX_UINT256),
+        userData: WeightedPoolEncoder.joinExactTokensInForBPTOut(ethDaiAmountsIn, 0),
+        fromInternalBalance: true, // Since we're joining from internal balance, we don't need to grant token allowance
+      },
+      0, // No eth
+      toChainedReference(17), // Store a reference for later staking
+    ]);
+
+    const stakeCalldata = library.interface.encodeFunctionData('gaugeDeposit', [
+      ETH_DAI_GAUGE,
+      relayer.address,
+      sender.address,
+      toChainedReference(17), // Stake all BPT from the join
+    ]);
+
+    await relayer.connect(sender).multicall([unstakeCalldata, exitCalldata, joinCalldata, stakeCalldata]);
+
+    expect(await destinationGauge.balanceOf(sender.address)).to.be.gt(0);
+  });
+
+  it('sender can update their gauge liquidity limits', async () => {
+    const tx = await relayer.connect(sender).multicall([
+      library.interface.encodeFunctionData('manageUserBalance', [
+        [RETH_STABLE_GAUGE, ETH_STETH_GAUGE].map((gauge) => ({
+          kind: TRANSFER_EXTERNAL_USER_BALANCE_OP_KIND,
+          asset: gauge,
+          amount: 1, // The amount must be non-zero to trigger the token transfer and liquidity limit update
+          sender: sender.address,
+          recipient: sender.address,
+        })),
+        0, // value: 0 (no need to transfer ETH)
+        [], // no output references
+      ]),
+    ]);
```

Comment:
This test could use some rework. We have a helper now that does this in `GaugeActions` (`gaugeCheckpoint`) which does this but it's much simpler to call. It should work both in mainnet and the L2s.

Reply:
Did the mainnet one first.